lib/nrf_modem_lib/trace_backends/uart/uart.c

```diff
@@ -1,141 +1,159 @@
 
-/* Maximum time to wait for a UART transfer to complete before giving up. */
-#define UART_TX_WAIT_TIME_MS 100
-#define UNUSED_FLAGS 0
+static const struct device *const uart_dev = DEVICE_DT_GET(UART_DEVICE_NODE);
 
-/* Semaphore used to check if the last UART transfer was completed. */
+/* Maximum time to wait for a UART transfer to complete before aborting. */
+#define UART_TX_WAIT_TIME_MS 1000
+/* Maximum UART transfer attempts. */
+#define UART_TX_RETRIES 5
+
+/* Semaphore used to synchronize UART transfers. */
 static K_SEM_DEFINE(tx_sem, 1, 1);
+/* Whether last UART transfer was successful. */
+static bool tx_succeeded;
 
+/* Callback to notify the trace library when trace data is processed. */
 static trace_backend_processed_cb trace_processed_callback;
 
-static void uarte_callback(nrfx_uarte_event_t const *event, void *p_context)
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
 {
-	__ASSERT(k_sem_count_get(&tx_sem) == 0,
-			"UART TX semaphore not in use");
-
-	if (event->type == NRFX_UARTE_EVT_ERROR) {
-		LOG_ERR("uarte error 0x%04x", event->data.error.error_mask);
+	ARG_UNUSED(dev);
+	ARG_UNUSED(user_data);
 
+	switch (evt->type) {
+	case UART_TX_DONE:
+		tx_succeeded = true;
 		k_sem_give(&tx_sem);
-	}
-
-	if (event->type == NRFX_UARTE_EVT_TX_DONE) {
+		break;
+	case UART_TX_ABORTED:
+		tx_succeeded = false;
 		k_sem_give(&tx_sem);
+		break;
+	default:
+		LOG_DBG("Unhandled UART event: %d", evt->type);
+		break;
 	}
 }
 
-static void wait_for_tx_done(void)
-{
-	/* Attempt to take the TX semaphore to stop the thread execution until
-	 * UART is done sending. Immediately give the semaphore when it becomes
-	 * available.
-	 */
-	if (k_sem_take(&tx_sem, K_MSEC(UART_TX_WAIT_TIME_MS)) != 0) {
-		LOG_WRN("UARTE TX not available");
-	}
-
-	k_sem_give(&tx_sem);
-}
-
 int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
 {
 	int err;
-	const nrfx_uarte_config_t config = {
-		.skip_gpio_cfg = true,
-		.skip_psel_cfg = true,
-		.hal_cfg.hwfc = NRF_UARTE_HWFC_DISABLED,
-		.hal_cfg.parity = NRF_UARTE_PARITY_EXCLUDED,
-		.baudrate = NRF_UARTE_BAUDRATE_1000000,
-		.interrupt_priority = DT_IRQ(UART1_NL, priority),
-		.p_context = NULL,
-	};
+	struct uart_config uart_cfg;
 
 	if (trace_processed_cb == NULL) {
 		return -EFAULT;
 	}
 
 	trace_processed_callback = trace_processed_cb;
 
-	err = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(UART1_NL), PINCTRL_STATE_DEFAULT);
-	__ASSERT_NO_MSG(err == 0);
+	if (!device_is_ready(uart_dev)) {
+		printk("UART device not found!");
+		return -ENODEV;
+	}
+
+	err = uart_config_get(uart_dev, &uart_cfg);
+	if (err) {
+		LOG_ERR("UART_config_get failed: %d", err);
+		return err;
+	}
 
-	IRQ_CONNECT(DT_IRQN(UART1_NL), DT_IRQ(UART1_NL, priority), nrfx_isr,
-		    &nrfx_uarte_1_irq_handler, UNUSED_FLAGS);
+	if (uart_cfg.baudrate < TRACE_UART_BAUD_RECOMMENDED) {
+		LOG_WRN("Trace UART baudrate is lower than recommended (%d<%d)",
```

Comment:
I think that lower baudrates do not work with Fidoless-tracing using trace collector. Should this be mentioned in the warning?

Reply:
Setting a different baud rate would likely work with FIDO-tracing, so I am a bit unsure what approach should be taken here.

Can someone offer an opinion on this?

Reply:
That is unrelated to the trace backends and modem tracing in NCS.

Reply:
Check. Had used FIDO tracing exactly once, and had only recollection that the baudrate could be controlled.